File: web/src/types/manga.ts
import { Cover, Meta, Filterable } from "@/types/general"
import { ReaderChapter, ReaderVolume, ReaderSettings } from "@/types/reader"

export interface Manga extends Filterable
{
    id: number
    // 
    name: string
    eng_name: string
    slug: string
    status: string
    // 
    mature: boolean
    // 
    cover: Cover
    covers: Cover[]
    // 
    ext: string
    filesize: number
    foldersize: number
    download_path: string
    meta: Meta
    timestamp: number
}

export interface MangaVolume
{
    id: number
    // 
    number: number
    name: string
    eng_name: string
    slug: string
    status: string
    // 
    cover: Cover
    // 
    ext: string
    filesize: number
    foldersize: number
    download_path: string
    meta: Meta
    // 
    chapters: MangaChapter[]
}

export interface MangaChapter
{
    id: number
    // 
    number: number
    name: string
    eng_name: string
    // 
    ext: string
    filesize: number
    download_path: string
}


export interface MangaSingle extends Manga
{
    volumes: MangaVolume[]
}

// 

export interface CachedMangaSingle
{
    data: MangaSingle
    timestamp: number
}

export interface CachedMangaList
{
    data: Manga[]
    timestamp: number
}

// 

export interface ReaderManga
{
    id: number
    name: string
    eng_name: string
    slug: string
    // 
    timestamp: number
    chapters: ReaderMangaChapter[]
}

export interface ReaderMangaVolume extends ReaderVolume
{
    chapters: ReaderMangaChapter[]
}

export interface ReaderMangaChapter extends ReaderChapter
{
    filesize: number
    download_path: string
}

export interface ReaderMangaSettings extends ReaderSettings {}

File: app/schemas/manga.py
from __future__ import annotations

from pydantic import BaseModel, Field
from datetime import date, datetime
from typing import List, Dict, Any, Union, Optional

from .core import *

class MangaMeta( BaseModel ):
    key: str = Field( validation_alias='meta_key' )
    value: str = Field( validation_alias='meta_value' )

    class Config:
        from_attributes = True

class Manga( BaseModel ):
    id: int
    #
    name: str = ""
    eng_name: str = ""
    slug: str = ""
    status: str = ""
    #
    authors: List[ Author ] = []
    genres: List[ Genre ] = []
    #
    mature: bool = False
    #
    cover: Cover | None
    covers: List[ Cover ] = Field( validation_alias='all_covers' )
    #
    # volumes_count: int = 0
    #
    ext: str = ""
    foldersize: int = 0
    meta: Dict[ str, Any ] = {}
    timestamp: int = 0

    class Config:
        from_attributes = True


class MangaVolume( BaseModel ):
    id: int
    manga_id: int
    #
    number: float = 1
    name: str = ""
    eng_name: str = ""
    slug: str = ""
    status: str = ""
    #
    cover: Cover | None
    #
    ext: str = ""
    filesize: int = 0
    foldersize: int = 0
    download_path: str = ""
    meta: Dict[ str, Any ] = {}

    class Config:
        from_attributes = True


class MangaChapter( BaseModel ):
    id: int
    volume_id: int
    volume_number: float
    #
    number: float = 1
    name: str = ""
    eng_name: str = ""

    ext: str = ""
    filesize: int = 0
    download_path: str = ""

    class Config:
        from_attributes = True


#

class MangaReader( BaseModel ):
    id: int
    name: str = ""
    eng_name: str = ""
    slug: str = ""
    timestamp: int = 0
    chapters: List[ MangaReaderChapter ] = []

    class Config:
        from_attributes = True


class MangaReaderChapter( BaseModel ):
    number: float = 1
    volume_number: float
    # 
    filesize: int = 0
    download_path: str = ""

    class Config:
        from_attributes = True
#


class MangaFull(Manga):
    volumes: List[ MangaVolumeFull ] = []

    class Config:
        from_attributes = True


class MangaVolumeFull(MangaVolume):
    chapters: List[ MangaChapter ] = []

    class Config:
        from_attributes = True